src/v/ssx/include/ssx/single_sharded.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/shard_id.hh>
+#include <seastar/core/sharded.hh>
+
+#include <limits>
+#include <optional>
+#include <utility>
+
+namespace ssx {
+
+/// Template helper to run a real service service on one core and an small dummy
+/// on other cores.
+///
+/// \tparam Service a class to be instantiated on each core.  Must expose
+///         a \c stop() method that returns a \c future<>, to be called when
+///         the service is stopped.
+template<typename Service>
+class maybe_service {
+    using underlying_t = std::optional<Service>;
+    underlying_t _underlying;
+
+public:
+    maybe_service() = default;
+
+    template<class... Args>
+    explicit maybe_service(bool start, Args&&... args) {
+        if (start) {
+            _underlying.emplace(std::forward<Args>(args)...);
+        }
+    }
+
+    ss::future<> stop() {
+        return _underlying ? _underlying->stop() : ss::make_ready_future();
+    }
+
+    // the ones below are only safe to be called on the shards where the service
+    // has started
+
+    constexpr const Service* operator->() const noexcept {
+        return &*_underlying;
+    }
+    constexpr Service* operator->() noexcept { return &*_underlying; }
+
+    constexpr const Service& operator*() const& noexcept {
+        return *_underlying;
+    }
+    constexpr Service& operator*() & noexcept { return *_underlying; }
+    constexpr const Service&& operator*() const&& noexcept {
+        return &*_underlying;
+    }
+    constexpr Service&& operator*() && noexcept { return &*_underlying; }
+};
+
+/// Template helper to run a real service service on one core and an small dummy
+/// on other cores.
+///
+/// \tparam Service a class to be instantiated on the core.  Must expose
+///         a \c stop() method that returns a \c future<>, to be called when
+///         the service is stopped.
+template<typename Service>
+class single_sharded : ss::sharded<maybe_service<Service>> {
```

Comment:
can we add a `local()` method here as well ? it may be convenient to use with the services running on the same shard (it can `vassert` if used on incorrect shard) ?

Reply:
vassert doesn't seem to be used in ssx, I assume that's for being able to merge ssx stuff into Seastar

so I just throw an exception, like normal sharded services do